src/main/scala/apps/cameraPipe.scala

```diff
@@ -0,0 +1,400 @@
+package apps
+
+import rise.core._
+import rise.core.types._
+import rise.core.DSL._
+import rise.core.TypeLevelDSL._
+import rise.core.HighLevelConstructs._
+
+// algorithm taken from Halide (halide-lang.org)
+object cameraPipe {
+  // iN -> iN -> uN
+  val abs_diff: Expr = foreignFun("abs_diff_i16", i16 ->: i16 ->: u16)
+  def clamp(dt: DataType): Expr =
+    foreignFun(s"clamp_${dt}", dt ->: dt ->: dt ->: dt)
+  val max: Expr = foreignFun("max_i16", i16 ->: i16 ->: i16)
+  val pow: Expr = foreignFun("pow_f32", f32 ->: f32 ->: f32)
+
+  def li16(v: Int): Expr = cast(l(v)) :: i16
+  def li32(v: Int): Expr = cast(l(v)) :: i32
+  def lu8(v: Int): Expr = cast(l(v)) :: u8
+
+  // average two positive values rounding up
+  val avg: Expr = implDT(dt => dtFun(compute_dt => fun(
+    dt ->: dt ->: dt
+  )((a, b) => {
+    val cast_ = fun(a => cast(a) :: compute_dt)
+    // FIXME: avoid literal casts?
+    cast((cast_(a) + cast_(b) + cast_(l(1))) / cast_(l(2))) :: dt
+  })))
+
+  case class Image(
+    xBeg: Nat, xEnd: Nat,
+    yBeg: Nat, yEnd: Nat,
+    expr: Expr
+  )
+
+  def letImage(img: Image, f: Image => Expr): Expr =
+    img.expr |> fun(x =>
+      f(Image(img.xBeg, img.xEnd, img.yBeg, img.yEnd, x))
+    )
+
+  def mapImage(img: Image, f: Expr): Image =
+    Image(img.xBeg, img.xEnd, img.yBeg, img.yEnd, img.expr |> map(map(f)))
+
+  def zipImage(a: Image, b: Image): Image =
+    // TODO: check a and b regions?
+    Image(a.xBeg, a.xEnd, a.yBeg, a.yEnd, zipND(2)(a.expr, b.expr))
```

Comment:
that is why there is the TODO comment. although the zip type will already check array sizes (but not image regions)

Reply:
ok, more general question: do you think it's beneficial or helpful to work with the `Image` abstraction because it's more similar to Halide? Phrased differently: would working with plain arrays make it harder to compare our stuff against halides?

Reply:
I am not sure what you are asking exactly here. I introduced the image abstraction because it automates keeping track of the valid domain of an image, and enables to automatically insert `drop`s and `take`s when intersecting/zipping multiple images. Dealing with this manually was a pain. Maybe there are other abstractions to simplify writing the algorithm, but this one was the most straightforward I could think of. Of course I was influenced by the Halide DSL that is my reference and I tried to have something as easy to compare as possible.

Reply:
A simple example of the added information with a 1D sliding window:
```
a : 4.T
a |> slide(2) : 3.2.T
i : [0-3].T
i |> stencilCollect(Seq(-1, 0)) : [1-3].2.T ~ 3.2.T
i |> stencilCollect(Seq(0, 1)) : [0-2].2.T ~ 3.2.T
```